hugr-core/src/export.rs

```diff
@@ -0,0 +1,416 @@
+//! Text syntax.
+use crate::{
+    extension::ExtensionSet,
+    ops::{NamedOp as _, OpType},
+    types::{
+        type_param::{TypeArgVariable, TypeParam},
+        type_row::TypeRowBase,
+        CustomType, FuncTypeBase, MaybeRV, PolyFuncTypeBase, RowVariable, SumType, TypeArg,
+        TypeBase, TypeEnum,
+    },
+    Hugr, HugrView, Node, Port,
+};
+use hugr_model::v0 as model;
+use indexmap::IndexSet;
+use smol_str::ToSmolStr;
+use tinyvec::TinyVec;
+
+/// Export a [`Hugr`] graph to its representation in the model.
+pub fn export(hugr: &Hugr) -> model::Module {
+    let mut context = Context::new(hugr);
+    hugr.root().export(&mut context);
+    context.module
+}
+
+/// State for converting a HUGR graph to its representation in the model.
+struct Context<'a> {
+    /// The HUGR graph to convert.
+    hugr: &'a Hugr,
+    /// The module that is being built.
+    module: model::Module,
+    /// Mapping from ports to edge indices.
+    /// This only includes the minimum port among groups of linked ports.
+    edges: IndexSet<(Node, Port)>,
+}
+
+impl<'a> Context<'a> {
+    pub fn new(hugr: &'a Hugr) -> Self {
+        Self {
+            hugr,
+            module: model::Module::default(),
+            edges: IndexSet::new(),
+        }
+    }
+
+    pub fn make_term(&mut self, term: model::Term) -> model::TermId {
+        let index = self.module.terms.len();
```

Comment:
Is this the place to do hash-consing?? (I.e. look for an existing `Term` that matches the suppiled, and if so then return that TermId rather than pushing)

Reply:
It will be useful to associate ids with uses, i.e be able to trace problems back to the source by figuring out which instance of a term we're looking at. I'm looking in to hashconsing at serialisation time, and preserving all the ids that point to the same term. We can apply this to the model separately without changing the model behaviour I think.

Reply:
have experimented with doing hashconsing here: e4ef7be6b2c68f30c5cb8826ba623ab07b8eab5c
I think it's worthwhile.